Adds a Classmap Generator

mikeschinkel opened the pull request:
This adds a classmap generator to the toolchain which generates a file `/wp-classmap.php` into the root.  This PR includes an example of that file so you can see what it will look like but of course it will be regenerated on build.

**NOTE**: I attempted to add the task to Grunt as `classmapgen` and include the task in the `build` task but since I do not currently have Grunt set up on my machine I could not test it. Hopefully it is trivially simple to fix?

schlessera commented:
@mikeschinkel You don't need to have grunt installed globally, you only need node.js with npm, which you probably have.

So, if you have `npm` available, you can just do `npm install && grunt`.

Also, your class map generation can be externally triggered through a simple HTTP request, which could prove very useful for a DOS attack. You might want to change this to not accept any request. Ideally, your class map generation would not be needed at run-time and could be moved outside of the `src` folder.

mikeschinkel commented:
> _You don't need to have grunt installed globally, you only need node.js with npm, which you probably have_

I did not say anything about installation, I said it was not _"set up"_.  That was short hand for _"It may take me some time to get Grunt working in my current VM environment even though it is installed because I do not normally use Grunt, someone else on our team has always set that up. And since it may take me some time I don't want to tackle that until I know I have the time I can reserve to accomplish it because working on this is not the only thing I have to do with my life."_

> Also, your class map generation can be externally triggered through a simple HTTP request, which could prove very useful for a DOS attack. You might want to change this to not accept any request.

Not true.  Because ...

>  Ideally, your class map generation would not be needed at run-time and could be moved outside of the src folder..

It is in the root of the repo, not in the `/src/` directory.

schlessera commented:
> I have Grunt installed, but I have not used it for a project yet so do not want to try to figure it out when I don't have enough time reserved in case it becomes a rabbit hole of required learning.

It's all set up already, as it is part of the build pipeline. As I said, you only need `npm`, the rest is automated.

> ...it is in the root of the repo, not in the /src/ directory.

Ah, yes, I see. I think you had it in the `src` folder in your own repo, so I just assumed it was the same here. Did not look closely enough! ;) Nevermind that point then.

mikeschinkel commented:
@schlessera 
> _It's all set up already, as it is part of the build pipeline. As I said, you only need npm, the rest is automated._

Put more bluntly, I don't have time to play with Grunt right now.  The POTUS debate is Monday and I have to complete something for a client.

GaryJones commented on src/wp-classmap.php:
```diff
@@ -0,0 +1,343 @@
+  'WP_Widget_Tag_Cloud' => 'wp-includes/widgets/class-wp-widget-tag-cloud.php',
+  'WP_Widget_Text' => 'wp-includes/widgets/class-wp-widget-text.php',
+  'wpdb' => 'wp-includes/wp-db.php',
+);
```
Clear line needed at the end of the file.

mikeschinkel replied:
Fixed

GaryJones commented on wp-classmap-generator.php:
```diff
@@ -0,0 +1,9 @@
+$generator->add_files( 'wp-admin' );
+$generator->add_files( 'wp-includes' );
+$classmap = $generator->get_classmap();
+file_put_contents( __DIR__ . '/src/wp-classmap.php', $classmap );
```
Clear trailing line needed here too.

mikeschinkel replied:
Fixed

mikeschinkel commented:
So I went ahead and added a simple autoloader in and I implemented code that inspects a constant `USE_COMPOSER_AUTOLOADER` to determine if is should use the Composer autoloader or the custom autoloader.  That will allow you to try both approaches.

schlessera commented on src/wp-includes/class-wp-autoloader.php:
```diff
@@ -0,0 +1,47 @@
+
+			spl_autoload_register( array( $this, 'load' ), true, true );
+
+			if ( is_file( ABSPATH . 'wp-classmap.php' ) ) {
```
You should include the `spl_autoload_register()` call within this condition only. Otherwise, each class will run through an invalid autoloader if the file could not be loaded.

mikeschinkel replied:
I intentionally meant it that when in anticipation of a `register_autoload_classes()` method that themes and plugins can use.

schlessera commented on tools/classmap/class-classmap-generator.php:
```diff
@@ -0,0 +1,128 @@
+		 */
+		$token = '([a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*)';
+
+		if ( preg_match_all( "/\n\s*(final|abstract)?\s*(class|interface|trait)\s+{$token}/", $php_code, $matches ) ) {
```
If I read the regex correctly, this will fail for namespaced classes/interfaces/traits. Not really a problem now, but might cause unexpected surprises later on.

schlessera replied:
This will also fail for more exotic cases (which might or might not be relevant for Core):

* `class /* some comment */ ClassName {}`
* `/* some coment */ class ClassName {}`
* `class sοмεμτƒςнαяαςτεяs {}`

mikeschinkel replied:
If this were a general purpose autoloader meant to handle any potential scenario &mdash; which is the situation for Composer &mdash; I would agree that we should address these. However, one of the big benefits of a custom autoloader such as this is that we can set the rules in core for how classes must be declared and in ensure none of those edge cases are a concern. 

In the latter 3 cases we can be sure to that core will not use any of those scenarios, and in the case of core choosing to namespace classes/interfaces and/or traits someone can easily modify the autoloader to handle those use-cases when and if WordPress chooses to adopt namespaces in core.

schlessera commented on tools/classmap/class-classmap-generator.php:
```diff
@@ -0,0 +1,128 @@
+	public function get_classmap() {
+
+		ob_start();
+		var_export( $this->_classmap );
```
No need for output buffering here, you can just do `$classmap = var_export( $this->_classmap, true );`.

The second argument tells var_export to return the result instead of echoing it.

mikeschinkel replied:
Good catch.  I have even used that in published code before, but for some reason I forgot about it this time.  Will submit a revised comment a few minutes after I post this comment.

schlessera commented on src/wp-includes/class-wp-autoloader.php:
```diff
@@ -0,0 +1,47 @@
+	 */
+	function __construct() {
+
+		if ( ! defined( 'DISABLE_CORE_AUTOLOADER' ) ) {
```
What if `DISABLE_CORE_AUTOLOADER=false` ?

mikeschinkel replied:
That has from an earlier version I committed but before I pushed. See [the current version](/staylor/develop.wordpress/pull/2/files#diff-748e02cd742247bfc3117214bebb22e1R5).

Actually, be sure to read and comment [on the collective files changed](/staylor/develop.wordpress/pull/2/files) as opposed to looking at individual changesets because you might well comment on something that is moot, such as this.

schlessera replied:
This seems to be the current version (as I had looked at the collective files changed). You check for existence of the constant, but not what value it was set to.

mikeschinkel replied:
Please look again.  The state of the code when you commented use the following `if()` statement:

    if ( defined( 'USE_COMPOSER_AUTOLOADER' ) && USE_COMPOSER_AUTOLOADER  ) {

mikeschinkel replied:
My bad, you are correct. I was looking at other parts of the code.  Will fix next.

mikeschinkel replied:
Fixed.
